### PR TITLE
sa default # 49

### PR DESCRIPTION
**Pull request by vir-mir**
Hi!
I offer a patch for the default fields.

Do you think it makes sense for life?

@asvetlov 

Issues https://github.com/aio-libs/aiopg/issues/49

### CODE REVIEW

**Comment by asvetlov** [comment]
Wow! Very impressive.
Let me sleep on it and make careful review.

**Comment by vir-mir** [comment]
- Combed code.
- Add processing scalar, sequence and fix set value None where default not null.

**Comment by vir-mir** [comment]
@asvetlov any news?

**Comment by pomidoroshev** [comment]
@asvetlov Чё как?

**Comment by Eduard90** [comment]
We need default values!

File: aiopg/sa/engine.py [hunk 1]
```diff
@@ -2,20 +2,53 @@
...
+        for key, default in default_fields:
+            pd[key] = (yield from self._exec_default(cursor, default))
+
+        return pd
+
+    @asyncio.coroutine
+    def _exec_default(self, cursor, default):
+        if default.is_sequence:
+            sql = "select nextval('{}')".format(
+                self.dialect.identifier_preparer.format_sequence(default))
```
Comment (vir-mir) [inline aiopg/sa/engine.py]: sqlalchemy makes a queries to the db.
Of course, we can make queries in advance...
such behavior we should document it explicitly

**Comment by jettify** [comment]
@vir-mir thank you very much for this PR! Lets give @asvetlov chance to review on weekends, otherwise I will merge it on Monday.

**Comment by asvetlov** [comment]
Looks good.
@jettify please merge

**Comment by serg666** [comment]
Guys, what about this [comment](https://github.com/aio-libs/aiopg/issues/49#issuecomment-263150311) ?

**Comment by Eduard90** [comment]
I have problems with UUID (PostgreSQL) column type:
```
  File "/usr/lib/python3.5/uuid.py", line 137, in __init__
    hex = hex.replace('urn:', '').replace('uuid:', '')
AttributeError: 'UUID' object has no attribute 'replace'
```

Table:
```
InvoiceTable = NewTable('invoice', metadata,
                        Column('id', UUID, primary_key=True),
                        Column('affiliate_id', UUID),
                        Column('status', String, index=True, default=StatusEnum.not_approved.name),
                        Column('created_at', DateTime, default=datetime.now),
                        Column('deleted', Boolean, default=False),
                        Column('is_paid', Boolean, default=False),
                        )
```

Primary key (id) is UUID and "by default" it must generate new UUID on insert.
I have "monkey patch" for _execute method (class SAConnection) but it's very bad code ))

**Comment by vir-mir** [comment]
@Eduard90 

http://docs.sqlalchemy.org/en/latest/core/custom_types.html#backend-agnostic-guid-type
```python
class UUID(sa.types.TypeDecorator):
    impl = sa.String

    def __init__(self, length=255, *args, **kwargs):
        super().__init__(length=255, *args, **kwargs)

    def process_bind_param(self, value, dialect=None):
        if value and isinstance(value, uuid.UUID):
            return value.bytes
        elif value and not isinstance(value, uuid.UUID):
            raise ValueError('value %s is not a valid uuid.UUID' % value)
        else:
            return None

    def process_result_value(self, value, dialect=None):
        if value:
            return uuid.UUID(bytes=value)
        else:
            return None

    def is_mutable(self):
        return False

meta = sa.MetaData()
tbl = sa.Table('sa_tbl4', meta,
               sa.Column('id', UUID(), primary_key=True, default=uuid.uuid4))

...

yield from conn.execute(tbl.insert().values())
yield from conn.execute(tbl.insert().values())
yield from conn.execute(tbl.insert().values())
```

**Comment by jettify** [comment]
@vir-mir thanks! should we put this snippet to the FAQ?

**Comment by vir-mir** [comment]
@serg666 https://github.com/aio-libs/aiopg/pull/56#issuecomment-95560109

**Comment by vir-mir** [comment]
@jettify all say that it is necessary to use an approach TypeDecorator

**Comment by Eduard90** [comment]
@vir-mir Yes, i know about this way. But ... with this custom type Alembic generate not correct migrations.
This problem in Alembic maybe. But i want use UUID and migrations now )))

And sqlalchemy has "builtin" type:
`from sqlalchemy.dialects.postgresql import UUID`
I think it's more correct than custom type. Or not?

**Comment by Eduard90** [comment]
@vir-mir How i can talk with you directly? Skype, Telegram, etc?

**Comment by vir-mir** [comment]
This is problem alembic
```bash
  File "../site-packages/sqlalchemy/sql/schema.py", line 79, in _init_items
    item._set_parent_with_dispatch(self)
AttributeError: 'UUID' object has no attribute '_set_parent_with_dispatch'
```
you can use
`from sqlalchemy.dialects.postgresql import UUID`

But to make it work you need to add a parameter `default`

 ```python
from sqlalchemy.dialects.postgresql import UUID
import uuid

meta = sa.MetaData()
tbl = sa.Table('sa_tbl4', meta,
               sa.Column('id', UUID(as_uuid=True), primary_key=True, default=uuid.uuid4))

...

yield from conn.execute(tbl.insert().values())
yield from conn.execute(tbl.insert().values())
yield from conn.execute(tbl.insert().values())
```

My skype `virmir49`

**Pull request merged**